### PR TITLE
autograd: double backwards function for binary_cross_entropy loss

### PR DESCRIPTION
**Pull request by anjali411**

### CODE REVIEW

**Comment by gchanan** [comment]
Some feedback on the non-code parts:

The title "autograd for binary_cross_entropy" could be improved.  binary_cross_entropy already supports the common-case of single backwards autograd, so it would be good to specify you are specifically referring to double backwards here.

File: test/common_nn.py [hunk 1]
```diff
@@ -295,7 +295,7 @@ def bceloss_no_reduce_test():
             lambda i: F.binary_cross_entropy(i, t.type_as(i), reduction='none')),
         input_fn=lambda: torch.rand(15, 10).clamp_(2.8e-2, 1 - 2.8e-2),
         reference_fn=lambda i, *_: -(t * i.log() + (1 - t) * (1 - i).log()),
-        check_gradgrad=False,
+        check_gradgrad=True,
```
Comment (gchanan) [inline test/common_nn.py]: nit: `check_gradgrad` is the default, so you can just delete this line and the ones below.

File: tools/autograd/templates/Functions.cpp [hunk 2]
```diff
@@ -886,6 +886,41 @@ Tensor log_softmax_double_backward(const Tensor & grad, const Tensor & grad_outp
   return z * grad_output.sum(dim, true) * ((grad * z).sum(dim, true) - grad);
 }
 
+Tensor binary_cross_entropy_double_backward(const Tensor & grad_output, const Tensor & grad, const Tensor & input, const Tensor & target, const Tensor& weight, int64_t reduction) {
+  auto eps = 1e-12;
+  auto num = input * input - 2 * input * target + target;
+  auto denom = (input + eps) * (input + eps) * (1 - input + eps) * (1 - input + eps);
+  auto output = num/denom;
```
Comment (gchanan) [inline tools/autograd/templates/Functions.cpp]: nit: try to be consistent about spacing around operators.  Most of the time here you have spaces around the operators (which matches the usual convention, but here you don't.

File: tools/autograd/templates/Functions.cpp [hunk 3]
```diff
@@ -886,6 +886,41 @@ Tensor log_softmax_double_backward(const Tensor & grad, const Tensor & grad_outp
...
+  output *=grad;
```
Comment (gchanan) [inline tools/autograd/templates/Functions.cpp]: here too.

File: tools/autograd/templates/Functions.cpp [hunk 4]
```diff
@@ -886,6 +886,41 @@ Tensor log_softmax_double_backward(const Tensor & grad, const Tensor & grad_outp
...
+  output *= grad_output;
+  if (weight.defined()) {
+    output *= weight;
+  }
+  if (reduction == Reduction::Mean) {
+    return output/input.numel();
```
Comment (gchanan) [inline tools/autograd/templates/Functions.cpp]: here too.

File: tools/autograd/templates/Functions.cpp [hunk 5]
```diff
@@ -886,6 +886,41 @@ Tensor log_softmax_double_backward(const Tensor & grad, const Tensor & grad_outp
...
+  auto eps = 1e-12;
+  auto num = input - target;
+  auto denom = (input + eps) * (1 - input + eps);
+  auto output = num/denom;
+  output *= grad_output;
+  if (weight.defined()) {
+    output *= weight;
+  }
+  if (reduction == Reduction::Mean) {
+    return output/input.numel();
```
Comment (gchanan) [inline tools/autograd/templates/Functions.cpp]: and here.

File: tools/autograd/derivatives.yaml [hunk 6]
```diff
@@ -951,6 +951,10 @@
 - name: binary_cross_entropy(Tensor self, Tensor target, Tensor? weight=None, int reduction=Mean) -> Tensor
   self: binary_cross_entropy_backward(grad, self, target, weight, reduction)
 
+- name: binary_cross_entropy_backward(Tensor grad_output, Tensor self, Tensor target, Tensor? weight=None, int reduction=Mean) -> Tensor
+  self: binary_cross_entropy_double_backward(grad_output, grad, self, target, weight, reduction)
+  grad_output: binary_cross_entropy_double_backward_grad_output(grad_output, self, target, weight, reduction)
```
Comment (gchanan) [inline tools/autograd/derivatives.yaml]: this doesn't look correct -- for a pointwise function, you probably don't need the original grad_output (because you are taking the derivative wrt to it, so that term is just going to fall out.

I also think you could probably reuse the computation for the single-backward here if you want, with just passing in a tensor of 1s instead of the actual grad_output.

**Review by gchanan — CHANGES_REQUESTED** [review]

File: tools/autograd/templates/Functions.cpp [hunk 7]
```diff
@@ -902,6 +902,40 @@ Tensor log_softmax_double_backward(const Tensor & grad, const Tensor & grad_outp
   return z * grad_output.sum(dim, true) * ((grad * z).sum(dim, true) - grad);
 }
 
+Tensor binary_cross_entropy_double_backward(const Tensor & grad_output, const Tensor & grad, const Tensor & input, const Tensor & target, const Tensor& weight, int64_t reduction) {
+  auto eps = 1e-12;
```
Comment (albanD) [inline tools/autograd/templates/Functions.cpp]: I would have two questions here:
- Why `1e-12`?
- What is the type of eps? Are we sure it's not going to become `0`?

File: tools/autograd/templates/Functions.cpp [hunk 8]
```diff
@@ -902,6 +902,40 @@ Tensor log_softmax_double_backward(const Tensor & grad, const Tensor & grad_outp
...
+  auto num = input * input - 2 * input * target + target;
+  auto denom = (input + eps) * (input + eps) * (1 - input + eps) * (1 - input + eps);
```
Comment (albanD) [inline tools/autograd/templates/Functions.cpp]: Every time to do `input + eps` here, you actually create a new Tensor of the size of `input`, and add element wise `eps` to it.
You would speed this up and use less memory by doing (with different naming):
```cpp
auto input_plus_eps = input + eps;
auto one_minus_input_plus_eps = 1 - input_plus_eps;
auto denom = input_plus_eps.pow(2) * one_minus_input_plus_eps.pow(2);
```

File: tools/autograd/templates/Functions.cpp [hunk 9]
```diff
@@ -902,6 +902,40 @@ Tensor log_softmax_double_backward(const Tensor & grad, const Tensor & grad_outp
...
+  auto output = num / denom;
+  output *= (grad * grad_output);
```
Comment (albanD) [inline tools/autograd/templates/Functions.cpp]: I think this would be better as: `output.mul_(grad).mul_(grad_output);`.
That way you do not create an intermediary Tensor of the size of the output.

**Review by albanD — CHANGES_REQUESTED** [review]
Overall the code looks good. I agree with your formulas.
I have some comment on speed/memory optimization only.

Some high level comments:
- You should try and reuse computations as much as possible to avoid doing the same thing repeatedly. I gave you one example, but you can use this more broadly in the PR.
- Using inplace operation can be tricky but very useful. In particular, you can:
  - Add inplace operations everywhere as long as you don't change the inputs. For example, `auto output = num.div_(denom);`.
  - Add a test that check that your code can be differentiated through without raising "A variable needed to compute the gradient was modified inplace".
  - Remove the inplace operations that make this check fail.

File: tools/autograd/templates/Functions.cpp (hunk 7 above)
Comment (anjali411) [reply tools/autograd/templates/Functions.cpp]: Hey! 
1) I saw this epsilon value being used for calculating the first differential in the backward function, which is what I am doing here too so I just used the same value. 
2) backward function works fine with that eps, so I am assuming not?

**Review by albanD — APPROVED** [review]

File: tools/autograd/templates/Functions.cpp [hunk 10]
```diff
@@ -902,6 +902,40 @@ Tensor log_softmax_double_backward(const Tensor & grad, const Tensor & grad_outp
   return z * grad_output.sum(dim, true) * ((grad * z).sum(dim, true) - grad);
 }
 
+Tensor binary_cross_entropy_double_backward(const Tensor & grad_output, const Tensor & grad, const Tensor & input, const Tensor & target, const Tensor& weight, int64_t reduction) {
+  auto eps = 1e-12;
+  auto inp_pl_eps = input + eps;
+  auto one_m_inp_pl_eps = 1 - input + eps;
+  auto output = (input * input - 2 * input * target + target) / (inp_pl_eps.pow(2) * one_m_inp_pl_eps.pow(2));
+  output *= (grad * grad_output);
```
Comment (gchanan) [inline tools/autograd/templates/Functions.cpp]: I don't think output is a very good name -- it can be confused with the output of the original function.

What about something like ggI and ggO for grad_grad_input and grad_grad_output (below)?

**Review by gchanan — APPROVED** [review]
approved subject to the minor renaming.
